various fixes and tweaks for gitlab

meramsey opened the pull request:
Some more minor fixes and tweaks from using this amazing tool of yours. I wanted to share upstream.

Some other thoughts.

duties check-docs will always fail as it never has the coverage report in the way that things are setup. 

  INFO    -  Building documentation to directory: /home/onessa/PycharmProjects/domaininfowizard/site 
  WARNING -  mkdocs_coverage.plugin: No such HTML report directory: build/coverage 

  Exited with 1 warnings in strict mode.
  
  
It requires running this in order so that the coverage files exist when it tries to check/build docs in strict mode.
poetry run duty test
poetry run duty coverage
poetry run duty check-docs

Figured I'd mention that as it was pretty frustrating that that kept causing failures till realized.

whattheserver committed:
Update pytest.ini

whattheserver committed:
Update duties.py

Added detection for github and gitlab so documentation deployment can be adjusted accordingly.

Also added docs_deploy_gitlab function to duties

Reference:
https://gitlab.com/pages/mkdocs
https://squidfunk.github.io/mkdocs-material/publishing-your-site/#gitlab-pages
https://docs.gitlab.com/ee/user/project/pages/getting_started_part_one.html
https://medium.com/faun/automate-you-documentation-with-gitlab-and-mkdocs-26e1b564f9f2

whattheserver committed:
Update .gitlab-ci.yml

Updated to add python 3.9 to tests and to also setup gitlab pages on deploy stage.

whattheserver committed:
Merge pull request #1 from whattheserver/patch-2

Update .gitlab-ci.yml

whattheserver committed:
Update .gitlab-ci.yml

Updated to fix failing stuff due check-docs needing coverage to build successfully which is lost 

https://gitlab.com/mikeramsey/wizard-domaininfo/-/jobs/1083611766

whattheserver committed:
Update .gitlab-ci.yml

whattheserver committed:
Update README.md.jinja

Fixed hardcoded github.io even when gitlab is selected.

Set's this for gitlab documentation pages
gitlab.io

meramsey commented:
Got it to pass finally with the latest code :)
https://gitlab.com/mikeramsey/wizard-domaininfo/-/pipelines/267666884

Pages now working:
https://mikeramsey.gitlab.io/wizard-domaininfo/

whattheserver committed:
Add files via upload

Added precommit hook so that all tests are run via poetry run duty test.

https://pre-commit.com/#intro

 git commit -a
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check Yaml...............................................................Passed
Check for added large files..............................................Passed
black....................................................................Passed
mypy.....................................................................Passed
Check pytest unit tests pass.............................................Passed
[master 8da40c0] tests: Added new test assertions and minor fixes for coverage
 12 files changed, 43 insertions(+), 13 deletions(-)

This should be put in root of repo and then precommit installed
pip install pre-commit

pre-commit install

meramsey commented:
One more thing I was thinking was that for gitlab you can get automatic coverage report from the pytest coverage report if its output in the log.

https://docs.gitlab.com/ee/ci/pipelines/settings.html#test-coverage-parsing

For coverage on project automatically needs this in the projects settings
`^TOTAL.+?(\d+\.\d+\%)$`

Although it looks like we could just add this line to the test stage which has the output for tests in the .gitlab-ci.yml and it would work without people having to hunt down and set that preference.
`coverage: '^TOTAL.+?(\d+\.\d+\%)$'`

If your open to that change for gitlab i'd recommend it but its not in this MR.

pawamoy commented on project/.gitlab-ci.yml:
```diff
@@ -27,7 +28,10 @@ stages:
-  script: poetry run duty check-docs
+  script:
+    - poetry run duty test
+    - poetry run duty coverage
```
The goal here is just to check docs. We don't really care about the actual coverage, so instead you can just do this:

```
- mkdir -p build/coverage
- touch build/coverage/index.html
- poetry run duty check-docs
```

pawamoy commented on project/.gitlab-ci.yml:
```diff
@@ -59,3 +63,25 @@ python3.7:
+
+pages:
+  stage: deploy
+  image: python:3.6
```
I think python:3.8 would be preferable :slightly_smiling_face: 
Not strong on this though.

pawamoy commented on project/.gitlab-ci.yml:
```diff
@@ -59,3 +63,25 @@ python3.7:
+    - poetry --version
+    - poetry install -vv
+  script:
+    - poetry run duty test
+    - poetry run duty coverage
+    - poetry run duty docs-deploy-gitlab
```
This time, instead of re-running the tests, we should bring the coverage report from the test job, as an artifact. Not sure how to do this in GitHub Workflows, I will search again and report back.

meramsey replied:
I think i have this all setup properly now in latest version let me know what you think.

pawamoy commented on project/.pre-commit-config.yaml:
```diff
@@ -0,0 +1,28 @@
+# See https://pre-commit.com for more information
+# See https://pre-commit.com/hooks.html for more hooks
```
Sorry but I'm simply not sold on pre-commit :sweat_smile: 
I find it slow and redundant: it creates virtualenvs and installs tools in a cache somewhere, instead of reusing the one I already dev-depend on through Poetry.

I could consider adding support for it through a Copier option though, which would default to false. If the value is false, this file would then be deleted by the post generation hook.

If we go this way, note that I won't maintain it myself!

meramsey replied:
I can't argue with that. :) 

Optional but not enabled by default seems like a safe bet. 

I will say that it takes a bit to figure out and setup and its handy to know my stuff is passing before i push to CI which if i did too much could cost me money after too many times. 

Definitely don't expect you to maintain it but figured others might find it handy and can optionally enable it easily if desired.

pawamoy commented on project/config/pytest.ini:
```diff
@@ -13,3 +13,6 @@ addopts =
   --cov-config config/coverage.ini
+testpaths =
+    tests
+    integration
```
What's that for? You split your tests into two folders? Maybe you can convince me I should too? Otherwise I'll ask you to revert this!

meramsey replied:
I removed the integration folder as I'm not setup for that yet. but i did add this so its explicit as that's how its setup already.
https://docs.pytest.org/en/stable/customize.html#pytest-ini

pawamoy commented on project/duties.py:
```diff
@@ -20,6 +20,17 @@
 PTY = not WINDOWS and not CI
 
 
+# Detect which CI were using so we can adjust things as needed for mkdocs from site to public for Gitlab Pages
+# https://docs.gitlab.com/ee/ci/variables/predefined_variables.html
+# Available for all jobs executed in CI/CD. true when available.
+GITLAB_CI = os.environ.get("GITLAB_CI")
+
+# https://docs.github.com/en/actions/reference/environment-variables Always set to true when GitHub Actions is
+# running the workflow. You can use this variable to differentiate when tests are being run locally or by GitHub
+# Actions.
+GITHUB_CI = os.environ.get("GITHUB_ACTIONS")
```
This is nice, but I would prefer to have this logic *templated* instead of executed at runtime. I think few users use both CIs so it's not useful to have both variables in duties.py. Also, I think both GitHub and GitLab's CI define the `CI` environment variable, so these two variables are redundant with the `CI` one (see line 18).

pawamoy commented on project/duties.py:
```diff
@@ -263,6 +275,18 @@ def docs_deploy(ctx):
     ctx.run("mkdocs gh-deploy", title="Deploying documentation")
 
+ 
+@duty
+def docs_deploy_gitlab(ctx):
+    """
+    Deploy the documentation on Gitlab Pages.
+
+    Arguments:
+        ctx: The context instance (passed automatically).
+    """
+    ctx.run("mkdocs build -s --site-dir public", title="Deploying documentation")
```
Same comment as for the CI variables: it should be templated instead. It's not useful to have two actions to deploy to GitHub *and* GitLab pages. Users deploy to only one.

In that case, I would even argue that the site directory should be configured in `mkdocs.yml` itself instead of here!

meramsey replied:
Put in PR for that here.
https://github.com/pawamoy/copier-poetry/pull/12

Awesome simple elegant solution :)

meramsey replied:
I'd be happy to have this removed if we generalize docs_deploy from being Github specific
https://github.com/pawamoy/copier-poetry/blob/b5f0fb3210cbe6f91ac4d5fc65e3261613b96126/project/duties.py#L259

I think with https://github.com/pawamoy/copier-poetry/pull/12 and the latest tweaks in .gitlab-ci.yml it will work like you suggest and be least amount of bloat :)

pawamoy commented on project/duties.py:
```diff
@@ -299,7 +323,10 @@ def release(ctx, version):
         ctx.run("poetry build", title="Building dist/wheel", pty=PTY)
         ctx.run("poetry publish", title="Publishing version", pty=PTY)
-        ctx.run("mkdocs gh-deploy", title="Deploying documentation", pty=PTY)
+        if GITHUB_CI:
+            ctx.run("mkdocs gh-deploy", title="Deploying documentation", pty=PTY)
+        elif GITLAB_CI:
+            ctx.run("mkdocs build -s --site-dir public", title="Deploying documentation")
```
Again, this should be templated instead of decided at runtime.
Also, I think this can be simplified by directly calling the `docs_deploy` action:

```python
docs_deploy.run()  # use its own context
# or docs_deploy(ctx)  # use the current context
```

(possible since duty v0.6.0)

meramsey replied:
removed and put this which with above tweak in mkdocs.yml will build to proper path without specifying it in the command.

ctx.run("mkdocs build", title="Deploying documentation")

pawamoy reviewed (changes requested):
Nice additions, however I really favor having things templated to avoid cluttering generated projects with code that is not used! See my individual comments :slightly_smiling_face:

pawamoy commented:
> For coverage on project automatically needs this in the projects settings
`^TOTAL.+?(\d+\.\d+\%)$`

Sure! Feel free to add this to `.gitlab-ci.yml`!
However note that this will only give you a badge, whereas the docs are actually adding the coverage HTML report to the docs pages :slightly_smiling_face: 

> Figured I'd mention that as it was pretty frustrating that that kept causing failures till realized.

Yes, sorry about that :sweat_smile: 
I recently added that feature and didn't update CI accordinly :confused:

whattheserver committed:
Update mkdocs.yml.jinja

whattheserver committed:
Update .gitlab-ci.yml

Setup tests for python 3.8 to cache the build/coverage artifact for pages publish

python3.8:
  <<: *test
  image: python:3.8
  artifacts:
    paths:
    - build/coverage

Setup coverage to be extracted when coverage is run 
https://docs.gitlab.com/ee/ci/yaml/README.html#coverage
coverage: '^TOTAL.+?(\d+\.\d+\%)$'

whattheserver committed:
Update duties.py

whattheserver committed:
Update pytest.ini

 I added tests as i wanted pytest and pycharm to default to the proper paths ecplicityly when selecting or guessing where tests should be
https://docs.pytest.org/en/stable/customize.html#pytest-ini

whattheserver committed:
Update duties.py

meramsey commented on project/mkdocs.yml.jinja:
```diff
@@ -3,6 +3,7 @@ site_description: "[[ project_description ]]"
 site_url: "https://[[ repository_namespace ]].github.io/[[ repository_name ]]"
 repo_url: "https://github.com/[[ repository_namespace ]]/[[ repository_name ]]"
 repo_name: "[[ repository_namespace ]]/[[ repository_name ]]"
+site_dir: '[% if repository_provider == "gitlab.com" %]public[% elif repository_provider == "github.com" %]site[% endif %]'
```
I believe this is the right format for this but if not feel free to edit it to be formatted properly for the jinja template. I used the readme.MD.jinja template as a reference.

whattheserver committed:
Update .gitlab-ci.yml

whattheserver committed:
Update .gitlab-ci.yml

Ended up seeing failures when passing build artifacts to deploy stage so while not ideal this seems to be the best combination till that issue is resolved.
 See https://gitlab.com/gitlab-org/gitlab/-/issues/14613 for reference.

Seems to run nicely now and only run coverage once when it does the last stage
https://gitlab.com/mikeramsey/wizard-domaininfo/-/jobs/1090053594

pawamoy committed:
refactor: Rework test coverage combining, adapt GitLab CI, clean up duties

pawamoy commented on project/.gitlab-ci.yml:
```diff
@@ -59,3 +66,27 @@ python3.7:
+  stage: deploy
+  <<: *install-deps
+  image: python:3.8
+  dependencies:
+  - python3.6
+  - python3.7
+  - python3.8
+  - python3.9
+  script:
+  - poetry run duty coverage
+  - poetry run duty docs
```
Retrieve artifacts from the listed jobs: the `.coverage-*` files. Combine them, generate the HTML report, build the docs.

pawamoy commented on project/.gitlab-ci.yml:
```diff
@@ -47,6 +51,9 @@ check-dependencies:
   <<: *install-deps
   stage: tests
   script: poetry run duty test
+  artifacts:
+    paths:
+    - .coverage-*
```
Keep generated coverage files. For Python 3.6, coverage file is named `.coverage-36`, etc.
See `duties.py`, actions `coverage` and `test`.

pawamoy commented on project/Makefile.jinja:
```diff
@@ -7,14 +7,16 @@ args = $(foreach a,$($(subst -,_,$1)_args),$(if $(value $a),$a="$($a)"))
 	clean \
 	coverage \
 	docs \
+	[%- if repository_provider == "github.com" %]
 	docs-deploy \
+	[%- endif %]
```
This action is not needed for GitLab. Docs are built and deployed from CI, using the `docs` action.

pawamoy commented on project/config/pytest.ini:
```diff
@@ -11,5 +11,6 @@ python_files =
   tests.py
 addopts =
   --cov
-  --cov-append
```
We don't need to append anymore. It was hackish anyway. For each Python version, the test run generates different coverage file. After every run, we can combine all coverage files into one. This is much cleaner!

pawamoy commented on project/duties.py.jinja:
```diff
@@ -253,6 +255,7 @@ def docs_serve(ctx, host="127.0.0.1", port=8000):
     ctx.run(f"mkdocs serve -a {host}:{port}", title="Serving documentation", capture=False)
 
 
+[% if repository_provider == "github.com" -%]
```
As stated previously, not needed when repository provider is GitLab.

pawamoy commented:
Hey @whattheserver, I just pushed some changes to your branch. Let me know what you think (I added comments to the code to explain) :wink:

meramsey commented:
> Hey @whattheserver, I just pushed some changes to your branch. Let me know what you think (I added comments to the code to explain) wink

I love them all. Do I need to do anything for this to be updated on my end? 

Or will this be handled when you merge it into yours as Master and i can just then reference that from going forward?

pawamoy commented:
When everything is perfect, I'll merge, and tag a new release. You'll then be able to `copier update` your proejcts :slightly_smiling_face: 

Note that I didn't test the GitLab stuff though, so maybe you should try it in one of your projects before we merge.

meramsey commented:
> When everything is perfect, I'll merge, and tag a new release. You'll then be able to `copier update` your proejcts slightly_smiling_face
> 
> Note that I didn't test the GitLab stuff though, so maybe you should try it in one of your projects before we merge.

I just modded the project i have on gitlab to match all the tweaks you did here and fired it off and it worked with the exception of using python3.8 with quality docs and artifacts and pages. 

So this all works perfect just recommend leaving those things to python 36 


Worked when on python 3.6 earlier
https://gitlab.com/mikeramsey/wizard-domaininfo/-/commit/d0a1512025849d280bac72344bb8cf9817835298

And again with just that changes from 3.8
https://gitlab.com/mikeramsey/wizard-domaininfo/-/pipelines/269230525


Doesn't when on python 3.8
https://gitlab.com/mikeramsey/wizard-domaininfo/-/jobs/1091566208
https://gitlab.com/mikeramsey/wizard-domaininfo/-/jobs/1091575985

Same versions for tests run fine and poety works fine though weird.

Can browse these results if curious seems like a gitlab issue with python 3.8 with artifacts

whattheserver committed:
Update .gitlab-ci.yml

Worked when on python 3.6 earlier
https://gitlab.com/mikeramsey/wizard-domaininfo/-/commit/d0a1512025849d280bac72344bb8cf9817835298

And again with just that changes from 3.8
https://gitlab.com/mikeramsey/wizard-domaininfo/-/pipelines/269230525


Doesn't when on python 3.8
https://gitlab.com/mikeramsey/wizard-domaininfo/-/jobs/1091566208
https://gitlab.com/mikeramsey/wizard-domaininfo/-/jobs/1091575985

Same versions for tests run fine and poety works fine though weird.

Can browse these results if curious seems like a gitlab issue with python 3.8 with artifacts

pawamoy commented:
I think it's a cache corruption issue. The Python 3.6 cached venv is restored while the job is now setup for Python 3.8, and it breaks somehow. We can leave it on Python 3.6 for now, it's fine :slightly_smiling_face:

pawamoy committed:
refactor: Fake coverage report directly in duty

This way it works both locally and in CI.

meramsey commented:
> I think it's a cache corruption issue. The Python 3.6 cached venv is restored while the job is now setup for Python 3.8, and it breaks somehow. We can leave it on Python 3.6 for now, it's fine slightly_smiling_face

Ah that makes sense.  Well is this safe to merge now?

I'm waiting to start a fresh skeleton for a new project and looking forward to testing out the next one and trying to splice fbs into poetry-copier and the duties for automating building
like this but for gitlab and since poetry would handle most of the dependencies it think this would be super handy
https://github.com/trappitsch/fbs-release-github-actions

pawamoy committed:
perf(ci): Reduce installation time in CI

Poetry caches artifacts and metadata into
its cache directory to speed things up
in subsequent installations.

Previously we were caching the virtualenv,
but Poetry was still taking a lot of time
to redownload the artifacts to build the
metadata.

Now we cache everything, so Poetry is very
fast both for locking
and installing dependencies.

pawamoy commented:
> Ah that makes sense. Well is this safe to merge now?

Yes.

I think it's good now. Still a few things to do but I don't want to push to your master branch more (next time you should definitely use another branch :smile:). I think I'll merge as is (with all 19 commits), but will squash most of it once in master, and force push. Not the best-practice but heh, if done fast, nobody will be harmed :stuck_out_tongue: 

Thanks again for your improvements!

pawamoy commented:
Done. Check out master branch. Let me know if there's anything wrong.
I'll tag a new release soon :rocket:

meramsey commented:
It looks great to me. Looking forward to it :)
